src/database.py

```
@@ -0,0 +1,98 @@
+import json
+from typing import List, Optional
+
+from sqlalchemy import Dialect, ForeignKey, Integer, SmallInteger, String
+from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship
+from sqlalchemy.types import TEXT, VARCHAR, TypeDecorator
+
+
+class JsonEncodeDict(TypeDecorator):
+    impl = VARCHAR
+
+    cache_ok = True
+
+    def process_bind_param(self, value: Optional[dict], dialect: Dialect):
+        return json.dumps(value) if value is not None else None
+
+    def process_result_value(self, value: Optional[str], dialect: Dialect):
+        return json.loads(value) if value is not None else None
+
+
+class GroupSet(TypeDecorator):
+    impl = TEXT
+
+    cache_ok = True
+
+    def process_bind_param(self, value: Optional[set], dialect: Dialect):
+        return ",".join(list(value)) if value else ""
+
+    def process_result_value(self, value: Optional[str], dialect: Dialect):
+        return set(value.split(",")) if value else set()
+
+
+class Base(DeclarativeBase):
```

this(the whole file) is great, although it's tightly connected to the dB schema provided by GLAuth

would be good to introduce a comment as a start saying we support version x y.z and then start to make a check on the database version or application version somehow (something we can ask to introduce upstream) 

basically we need to cover for the case three are breaking changes on the database schema and we crash due to that

---

I was also thinking about this, but I was not sure if we actually care. If a breaking migration happens, the sql query will fail and we can throw an error. It is cleaner what you suggest though, to set the charm status to blocked.

It is possible to verify the schema using sqlalchemy (something like https://stackoverflow.com/questions/30428639/check-database-schema-matches-sqlalchemy-models-on-application-startup), but it is not clear to me when that check should run (should every time an action is triggered? I think that even this would be acceptable given the fact that juju actions take too long to run). Your suggestion would be cheaper (performance wise), but it wouldn't catch any differences between our models and the actual schema. Differences between our models and the schema may happen because we manually create the models, but it would be fine if we are careful when we update our models.

---

Thanks for bringing this up. I will put it into a ticket as a note for further improvements. Somehow similar scenario to the Kratos and Hydra when upstream makes some breaking changes in the database schema.